File: tests/regression/900_EXAMPLES_SHOWCASE/910_LANGUAGE_SHOOTOUT/2101_nbody_optimized/reference/baseline.zig
// Hand-optimized Zig baseline for N-body simulation
// This is what Koru SHOULD compile to
//
// Based on: https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/nbody-gcc-1.html
// License: Revised BSD
// Ported to Zig: 2024-10-25
//
// Algorithm: Symplectic integration of planetary orbits
// - Calculate gravitational interactions between all pairs
// - Update velocities based on forces
// - Update positions based on velocities
// - Repeat N times

// OPTIMIZATION: Enable fast-math mode for aggressive FP optimization
// This allows FMA (fused multiply-add), reassociation, and other transformations
// Equivalent to C's -ffast-math or Rust's default behavior
comptime {
    @setFloatMode(.optimized);
}

const std = @import("std");

const PI = 3.141592653589793;
const SOLAR_MASS = 4 * PI * PI;
const DAYS_PER_YEAR = 365.24;

const Body = struct {
    x: f64,
    y: f64,
    z: f64,
    vx: f64,
    vy: f64,
    vz: f64,
    mass: f64,
};

fn advance(bodies: []Body, dt: f64) void {
    // Update velocities based on gravitational interactions
    var i: usize = 0;
    while (i < bodies.len) : (i += 1) {
        var j: usize = i + 1;
        while (j < bodies.len) : (j += 1) {
            const dx = bodies[i].x - bodies[j].x;
            const dy = bodies[i].y - bodies[j].y;
            const dz = bodies[i].z - bodies[j].z;
            const distance = @sqrt(dx * dx + dy * dy + dz * dz);
            const mag = dt / (distance * distance * distance);

            bodies[i].vx -= dx * bodies[j].mass * mag;
            bodies[i].vy -= dy * bodies[j].mass * mag;
            bodies[i].vz -= dz * bodies[j].mass * mag;

            bodies[j].vx += dx * bodies[i].mass * mag;
            bodies[j].vy += dy * bodies[i].mass * mag;
            bodies[j].vz += dz * bodies[i].mass * mag;
        }
    }

    // Update positions based on velocities
    for (bodies) |*body| {
        body.x += dt * body.vx;
        body.y += dt * body.vy;
        body.z += dt * body.vz;
    }
}

fn energy(bodies: []const Body) f64 {
    var e: f64 = 0.0;

    for (bodies, 0..) |body, i| {
        // Kinetic energy
        e += 0.5 * body.mass * (body.vx * body.vx + body.vy * body.vy + body.vz * body.vz);

        // Potential energy (pairwise)
        var j = i + 1;
        while (j < bodies.len) : (j += 1) {
            const dx = body.x - bodies[j].x;
            const dy = body.y - bodies[j].y;
            const dz = body.z - bodies[j].z;
            const distance = @sqrt(dx * dx + dy * dy + dz * dz);
            e -= (body.mass * bodies[j].mass) / distance;
        }
    }

    return e;
}

fn offsetMomentum(bodies: []Body) void {
    var px: f64 = 0.0;
    var py: f64 = 0.0;
    var pz: f64 = 0.0;

    for (bodies) |body| {
        px += body.vx * body.mass;
        py += body.vy * body.mass;
        pz += body.vz * body.mass;
    }

    bodies[0].vx = -px / SOLAR_MASS;
    bodies[0].vy = -py / SOLAR_MASS;
    bodies[0].vz = -pz / SOLAR_MASS;
}

pub fn main() !void {
    // Parse command line argument
    const args = try std.process.argsAlloc(std.heap.page_allocator);
    defer std.process.argsFree(std.heap.page_allocator, args);

    if (args.len < 2) {
        std.debug.print("Usage: {s} <iterations>\n", .{args[0]});
        return;
    }

    const n = try std.fmt.parseInt(u32, args[1], 10);

    // Initialize bodies (Sun, Jupiter, Saturn, Uranus, Neptune)
    var bodies = [_]Body{
        .{ // Sun
            .x = 0, .y = 0, .z = 0,
            .vx = 0, .vy = 0, .vz = 0,
            .mass = SOLAR_MASS,
        },
        .{ // Jupiter
            .x = 4.84143144246472090e+00,
            .y = -1.16032004402742839e+00,
            .z = -1.03622044471123109e-01,
            .vx = 1.66007664274403694e-03 * DAYS_PER_YEAR,
            .vy = 7.69901118419740425e-03 * DAYS_PER_YEAR,
            .vz = -6.90460016972063023e-05 * DAYS_PER_YEAR,
            .mass = 9.54791938424326609e-04 * SOLAR_MASS,
        },
        .{ // Saturn
            .x = 8.34336671824457987e+00,
            .y = 4.12479856412430479e+00,
            .z = -4.03523417114321381e-01,
            .vx = -2.76742510726862411e-03 * DAYS_PER_YEAR,
            .vy = 4.99852801234917238e-03 * DAYS_PER_YEAR,
            .vz = 2.30417297573763929e-05 * DAYS_PER_YEAR,
            .mass = 2.85885980666130812e-04 * SOLAR_MASS,
        },
        .{ // Uranus
            .x = 1.28943695621391310e+01,
            .y = -1.51111514016986312e+01,
            .z = -2.23307578892655734e-01,
            .vx = 2.96460137564761618e-03 * DAYS_PER_YEAR,
            .vy = 2.37847173959480950e-03 * DAYS_PER_YEAR,
            .vz = -2.96589568540237556e-05 * DAYS_PER_YEAR,
            .mass = 4.36624404335156298e-05 * SOLAR_MASS,
        },
        .{ // Neptune
            .x = 1.53796971148509165e+01,
            .y = -2.59193146099879641e+01,
            .z = 1.79258772950371181e-01,
            .vx = 2.68067772490389322e-03 * DAYS_PER_YEAR,
            .vy = 1.62824170038242295e-03 * DAYS_PER_YEAR,
            .vz = -9.51592254519715870e-05 * DAYS_PER_YEAR,
            .mass = 5.15138902046611451e-05 * SOLAR_MASS,
        },
    };

    // Offset momentum so sun is at rest
    offsetMomentum(&bodies);

    // Print initial energy
    std.debug.print("{d:.9}\n", .{energy(&bodies)});

    // Run simulation
    var i: u32 = 0;
    while (i < n) : (i += 1) {
        advance(&bodies, 0.01);
    }

    // Print final energy
    std.debug.print("{d:.9}\n", .{energy(&bodies)});
}

File: tests/regression/900_EXAMPLES_SHOWCASE/910_LANGUAGE_SHOOTOUT/_archive/2101_nbody/reference/baseline.py
#!/usr/bin/env python3
"""N-body simulation - idiomatic Python"""
import sys
import math

PI = 3.141592653589793
SOLAR_MASS = 4 * PI * PI
DAYS_PER_YEAR = 365.24

bodies = [
    # Sun
    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, SOLAR_MASS],
    # Jupiter
    [4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01,
     1.66007664274403694e-03 * DAYS_PER_YEAR, 7.69901118419740425e-03 * DAYS_PER_YEAR,
     -6.90460016972063023e-05 * DAYS_PER_YEAR, 9.54791938424326609e-04 * SOLAR_MASS],
    # Saturn
    [8.34336671824457987e+00, 4.12479856412430479e+00, -4.03523417114321381e-01,
     -2.76742510726862411e-03 * DAYS_PER_YEAR, 4.99852801234917238e-03 * DAYS_PER_YEAR,
     2.30417297573763929e-05 * DAYS_PER_YEAR, 2.85885980666130812e-04 * SOLAR_MASS],
    # Uranus
    [1.28943695621391310e+01, -1.51111514016986312e+01, -2.23307578892655734e-01,
     2.96460137564761618e-03 * DAYS_PER_YEAR, 2.37847173959480950e-03 * DAYS_PER_YEAR,
     -2.96589568540237556e-05 * DAYS_PER_YEAR, 4.36624404335156298e-05 * SOLAR_MASS],
    # Neptune
    [1.53796971148509165e+01, -2.59193146099879641e+01, 1.79258772950371181e-01,
     2.68067772490389322e-03 * DAYS_PER_YEAR, 1.62824170038242295e-03 * DAYS_PER_YEAR,
     -9.51592254519715870e-05 * DAYS_PER_YEAR, 5.15138902046611451e-05 * SOLAR_MASS],
]

def advance(bodies, dt):
    for i in range(len(bodies)):
        for j in range(i + 1, len(bodies)):
            dx = bodies[i][0] - bodies[j][0]
            dy = bodies[i][1] - bodies[j][1]
            dz = bodies[i][2] - bodies[j][2]
            dist = math.sqrt(dx*dx + dy*dy + dz*dz)
            mag = dt / (dist * dist * dist)
            bodies[i][3] -= dx * bodies[j][6] * mag
            bodies[i][4] -= dy * bodies[j][6] * mag
            bodies[i][5] -= dz * bodies[j][6] * mag
            bodies[j][3] += dx * bodies[i][6] * mag
            bodies[j][4] += dy * bodies[i][6] * mag
            bodies[j][5] += dz * bodies[i][6] * mag
    for body in bodies:
        body[0] += dt * body[3]
        body[1] += dt * body[4]
        body[2] += dt * body[5]

def energy(bodies):
    e = 0.0
    for i, body in enumerate(bodies):
        e += 0.5 * body[6] * (body[3]*body[3] + body[4]*body[4] + body[5]*body[5])
        for j in range(i + 1, len(bodies)):
            dx = body[0] - bodies[j][0]
            dy = body[1] - bodies[j][1]
            dz = body[2] - bodies[j][2]
            e -= (body[6] * bodies[j][6]) / math.sqrt(dx*dx + dy*dy + dz*dz)
    return e

def offset_momentum(bodies):
    px = py = pz = 0.0
    for body in bodies:
        px += body[3] * body[6]
        py += body[4] * body[6]
        pz += body[5] * body[6]
    bodies[0][3] = -px / SOLAR_MASS
    bodies[0][4] = -py / SOLAR_MASS
    bodies[0][5] = -pz / SOLAR_MASS

n = int(sys.argv[1]) if len(sys.argv) > 1 else 1000
offset_momentum(bodies)
print(f"{energy(bodies):.9f}")
for _ in range(n):
    advance(bodies, 0.01)
print(f"{energy(bodies):.9f}")
